common/cpure_ast_typeinfer.ml

```diff
@@ -0,0 +1,310 @@
+  end in
+  checker#visit_formula_annot (None) exp
+
+(** Propagate type information upward/sideways along the AST. 
+ The returned boolean indicates whether or not a synthesis rule was applied.*)
+let synthesize_types (exp : typ option formula_annot) : typ option formula_annot * bool =
+  let synth = object (self)
+    inherit [_] mapreduce as super
+    method visit_'a _ (x : typ option) = (x, false)
+    method zero = false
+    method plus = (||)
+
+    method! visit_exp_annot ctx exp =
+      match exp with
+      (* List introduction rule: if one element of a list has type T, the list has type (List T). *)
+      | (List (elements, loc), None) ->
+        let subexprs, subrule_applied = List.map (self#visit_exp_annot ctx) elements |> List.split in
+        let possible_list_typ = elements |> List.map (fun (_, typ) -> typ) |> List.find_opt Option.is_some |> Option.join in
+        ((List (subexprs, loc), possible_list_typ),
+          Option.is_some possible_list_typ || List.exists ((=) true) subrule_applied)
+      (* ListCons introduction rule:
+          - if a head has type T, its ListCons has type (List T).
+          - if a tail has type (List T), so does its ListCons *)
+      | (ListCons (head, tail, loc), None) ->
+          let head, head_applied = self#visit_exp_annot ctx head in
+          let tail, tail_applied = self#visit_exp_annot ctx tail in
+          let subrules_applied = head_applied || tail_applied in
+          begin match (typeof head, typeof tail) with
+          | (Some head_typ, _) ->
+            (ListCons (head, tail, loc), Some (Globals.List head_typ)), true
+          | (_, Some tail_typ) ->
+            (ListCons (head, tail, loc), Some tail_typ), true
+          | _ ->
+            (ListCons (head, tail, loc), None), subrules_applied
+          end
+      | _ -> super#visit_exp_annot ctx exp
+
+  end in
+  synth#visit_formula_annot () exp
+
+let lift_option_from_list (ls : 'a option list) : 'a list option =
+  let rec go ?(acc=[]) = function
+    | [] -> Some acc
+    | (Some x)::xs -> go ~acc:(x::acc) xs
+    | None::_ -> None
+  in
+  go ls |> Option.map List.rev
+
+let fill_missing_annotations (fill: unit -> 'a) (f : 'a option formula_annot) =
+  let go = object
+    inherit [_] map
+    method visit_'a _ (typ : typ option) = Option.value typ ~default:(fill ())
+  end in
+  go#visit_formula_annot () f
+
+let infer_cpure_types (f: Cpure.formula) : typ formula_annot =
+  (* Printf.printf "Given formula %s to infer\n" (Cpure.string_of_ls_pure_formula [f]); *)
+  let annotated = annotate_cpure_formula f in
+  let rec do_further_inference (exp : typ option formula_annot) : typ option formula_annot =
+    let exp, check_progress = check_types exp in
+    let exp, infer_progress = synthesize_types exp in
+    if check_progress || infer_progress
+    then do_further_inference exp
+    else exp
+  in
+  let result = do_further_inference annotated in
+  (fill_missing_annotations fresh_tvar result)
```

Comment:
I saw there were initially some unit tests here. Would it be worth keeping them around?

Reply:
The unit tests were previously for a HM-style unification algorithm, and it only tested that the type constraints were being resolved correctly. The current implementation is something more general, which assigns types to the AST as a whole. So there would need to be a new set of unit tests for the new implementation.

Reply:
Okay, I'm fine with not adding new tests. How does the current implementation differ from HM?

Reply:
It just goes up and down the AST to apply typing rules (e.g. if a list expression has type `List[T]`, its elements have type `T`) repeatedly until all possible rules have been applied. I chose to do it this way for more flexibility, and since TVars have already been unified at this point.